torch/autograd/graph.py

```diff
@@ -113,16 +113,16 @@ class save_on_cpu():
 
     """
     def __init__(self, pin_memory=False):
+        self.cuda_avail = torch.cuda.is_available()
+
         def pack_to_cpu(tensor):
             if not pin_memory:
                 return (tensor.device, tensor.cpu())
 
-            packed = torch.empty(
-                tensor.size(),
-                dtype=tensor.dtype,
-                layout=tensor.layout,
-                pin_memory=(torch.cuda.is_available() and not tensor.is_sparse))
-            packed.copy_(tensor)
+            packed = tensor.to("cpu", non_blocking=True)
+            if self.cuda_avail and not tensor.is_sparse:
```

Comment:
you don't need to separately pin memory, it will be automatically pinned with non-blocking copy:
```
In [1]: import torch

In [2]: a=torch.randn(4, device="cuda")

In [3]: b=a.to("cpu", non_blocking=True)

In [4]: b.is_pinned()
Out[4]: True
```
However, it also means that this is an asynchronous call, and `packed` won't contain the necessary data by the time this function returns. It will be the responsibility of the caller to synchronize before use (and to make sure `packed` doesn't go out of scope before it is actually used again)

Reply:
Thanks for the review @ngimel!

It looks like we could simply do something like 

```
packed = tensor.to("cpu", non_blocking=True)
```

which also avoids the 2 copies that @albanD mentioned as an issue. Although, the concern about change in semantics makes sense, and this would probably be a BC-breaking change in that aspect. Since it helps the performance on a few reproducible FSDP workloads, could we consider adding a flag to allow non-blocking copies? @albanD

Reply:
Also I was wondering whether the nonblocking semantics will cause issues in this particular case of `save_on_cpu`. Based on my understanding of how this is implemented in autograd, `packed` tensor as opposed to the input argument tensor will be saved by autograd. Although when this tensor is reloaded in `unpack` by autograd, we move it back to GPU (also nonblocking) so as long as `save_on_cpu` was previously being used with appropriate synchronization, this change should not incur any issues.